src/inc/til/rle.h

```diff
@@ -0,0 +1,1064 @@
+                    if (new_total > index)
+                    {
+                        run_pos = index - total;
+                        break;
+                    }
+
+                    total = new_total;
+                }
+
+                return { it, run_pos };
+            }
+
+        private:
+            It it;
+            const It end;
+            size_type run_pos = 0;
+            size_type total = 0;
+        };
+
+        basic_rle(container&& runs, size_type size) :
+            _runs(std::forward<container>(runs)),
+            _total_length(size)
+        {
+        }
+
+        void _compact()
+        {
+            auto it = _runs.begin();
+            const auto end = _runs.end();
+
+            if (it == end)
+            {
+                return;
+            }
+
+            for (auto ref = it; ++it != end; ref = it)
+            {
+                if (ref->value == it->value)
+                {
+                    ref->length += it->length;
+
+                    while (++it != end)
+                    {
+                        if (ref->value == it->value)
+                        {
+                            ref->length += it->length;
+                        }
+                        else
+                        {
+                            *++ref = std::move(*it);
+                        }
+                    }
+
+                    _runs.erase(++ref, end);
+                    return;
+                }
+            }
+        }
+
+        inline void _check_indices(size_type start_index, size_type& end_index)
+        {
+            if (end_index > _total_length)
+            {
+                end_index = _total_length;
+            }
+
+            // start_index and end_index must be inside the inclusive range [0, _total_length].
+            if (start_index > end_index)
+            {
+                throw std::out_of_range("start_index <= end_index");
+            }
+        }
+
+        // Replace the range [start_index, end_index) with replacements.
+        void _replace_unchecked(size_type start_index, size_type end_index, const gsl::span<const rle_type> replacements)
+        {
+            //
+            //
+            //
+            // MUST READ: How this function (mostly) works
+            // -------------------------------------------
+            //
+            // ## Overview
+            //
+            // Assuming this instance consists of:
+            //   _runs == {{1, 3}, {2, 3}, {3, 3}}
+            // Or shown in a more visual way:
+            //   1 1 1|2 2 2|3 3 3
+            //
+            // If we're called with:
+            //   _replace_unchecked(3, 6, {{4, 2}, {5, 2}})
+            // Or shown in a more visual way:
+            //   1 1 1|2 2 2|3 3 3
+            //       ^    ^         <-- the first ^ is "start_index" (inclusive) and the second "end_index" (exclusive)
+            //       1 1|4|2        <-- the "replacements"

```

Comment:
I don't totally understand -- I think these replacements are `4 4|5 5`

Reply:
(line 712)

Reply:
Line 716 is correct. I'm correcting line 712 to reflect that.